contracts/SpokePoolV3Periphery.sol

```diff
@@ -53,153 +84,153 @@ contract SpokePoolPeripheryProxy is Lockable, MultiCaller {
         SpokePoolV3Periphery.DepositData calldata depositData,
         uint256 deadline,
-        uint8 v,
-        bytes32 r,
-        bytes32 s
+        bytes calldata permitSignature
     ) external nonReentrant {
+        (bytes32 r, bytes32 s, uint8 v) = _deserializeSignature(permitSignature);
         // For permit transactions, we wrap the call in a try/catch block so that the transaction will continue even if the call to
         // permit fails. For example, this may be useful if the permit signature, which can be redeemed by anyone, is executed by somebody
         // other than this contract.
-        try swapToken.permit(msg.sender, address(this), swapTokenAmount, deadline, v, r, s) {} catch {}
-        _callSwapAndBridge(
-            IERC20(address(swapToken)),
-            acrossInputToken,
-            exchange,
-            routerCalldata,
-            swapTokenAmount,
-            minExpectedInputTokenAmount,
-            depositData
+        try
+            IERC20Permit(address(swapData.swapToken)).permit(
+                msg.sender,
+                address(this),
+                swapData.swapTokenAmount,
+                deadline,
+                v,
+                r,
+                s
+            )
+        {} catch {}
+        _callSwapAndBridge(acrossInputToken, swapData, depositData);
+    }
+
+    /**
+     * @notice Uses permit2 to transfer tokens from a user before swapping a token on this chain via specified router and submitting an Across deposit atomically.
+     * Caller can specify their slippage tolerance for the swap and Across deposit params.
+     * @dev This function assumes the caller has properly set an allowance for the permit2 contract on this network.
+     * @dev This function assumes that the amount of token to be swapped is equal to the amount of the token to be received from permit2.
+     * @param acrossInputToken Address of the token that will be bridged via Across as the inputToken.
+     * @param depositor The owner of the permit2 signature and depositor for the Across spoke pool.
+     * @param swapData Specifies the params we need to perform a swap on a generic exchange.
+     * @param depositData Specifies the Across deposit params we'll send after the swap.
+     * @param permit The permit data signed over by the owner.
+     * @param transferDetails The spender's requested transfer details for the permitted token.
+     * @param signature The permit2 signature to verify against the deposit data.
+     */
+    function swapAndBridgeWithPermit2(
+        IERC20 acrossInputToken,
+        address depositor,
+        SpokePoolV3Periphery.SwapData calldata swapData,
+        SpokePoolV3Periphery.DepositData calldata depositData,
+        IPermit2.PermitTransferFrom calldata permit,
+        IPermit2.SignatureTransferDetails calldata transferDetails,
+        bytes calldata signature
+    ) external nonReentrant {
+        bytes32 witness = keccak256(
+            abi.encode(
```

Comment:
This will be updated to a new ACROSS_SWAP_AND_DEPOSIT_DATA_TYPE in a follow-up PR (along with the other swapAndBridge functions) so that the user can't get destroyed if a malicious person modified the swapData to accept a low output amount (and thus a low inputAmount for the across input token).

Reply:
i'm not sure I understand this comment, shouldn't you include the types here if we need to get this to get permit2 to work?

Reply:
You're right. This will be added here. I don't think this would revert per se, but right now it is vulnerable to griefing, since the swapData is not part of the permit witness.

Reply:
Added here. I needed to refactor the structs since there are some overlaps between the output tokens in swaps and the input tokens in deposits. Also, since we need to include all the swap and deposit data in the permit2 witness, I had to move some arguments which were formerly not in the `SwapData/DepositData` struct into those structs so that they would be included in the witness (things like `inputToken/inputAmount`). This was done here https://github.com/across-protocol/contracts/pull/782/commits/e5473c57e07ab0a517ab11e93b819265a4e7ec5e